src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -161,5 +162,45 @@ async def gather_coordinates_for_jvm_lockfile(
     return ArtifactRequirements(requirements)
 
 
+@frozen_after_init
+@dataclass
+class ValidatedJvmToolLockfileRequest:
+
+    options_scope: str
+    artifact_inputs: FrozenOrderedSet[str]
+
+    def __init__(self, tool: JvmToolBase):
+        self.options_scope = tool.options_scope
+        self.artifact_inputs = FrozenOrderedSet(tool.artifact_inputs)
+        self.lockfile = tool.resolved_lockfile()
+
+    def __hash__(self):
+        return hash((self.options_scope, self.artifact_inputs, self.lockfile))
+
+
+@rule(desc="Validate JVM lockfile")
+async def validate_jvm_lockfile(
+    request: ValidatedJvmToolLockfileRequest,
+) -> CoursierResolvedLockfile:
+
+    lockfile = request.lockfile
+    requirements = await Get(
+        ArtifactRequirements,
+        GatherJvmCoordinatesRequest(
+            request.artifact_inputs, f"[{request.options_scope}].artifacts"
+        ),
+    )
```

Comment:
I'm trying to reason about if this should be requesting `JvmLockfileRequest` or not, and then looking at the `request.artifacts`. It seems wrong for us to be computing the two requirements in distinct ways. But I'm having a hard time following everything

Reply:
Yeah, I had some difficulty here. Let's discuss what a refactor would look like elsewhere.

Reply:
Yea, not a blocker, but this is odd. I _think_ that I might have caused some of the oddity though, because global `scalac_plugins` have a lockfile, but _don't_ have an associated subsystem. And so `ValidatedJvmToolLockfileRequest` cannot take a subsystem as an argument, because not all requests have one.